core/src/main/java/com/linecorp/armeria/client/encoding/HttpDecodingClient.java

```diff
@@ -83,7 +83,14 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Ex
             // Client specified encoding, so we don't do anything automatically.
             return delegate().execute(ctx, req);
         }
-        req.headers().set(HttpHeaderNames.ACCEPT_ENCODING, acceptEncodingHeader);
+
+        req = HttpRequest.of(req, req.headers().toBuilder()
+                                     .set(HttpHeaderNames.ACCEPT_ENCODING, acceptEncodingHeader)
+                                     .build());
+        if (ctx.request() instanceof HttpRequest) {
```

Comment:
Is it more appropriate to compare `ctx.request()` to the original `req`? This instanceof seems like magic (took a couple minutes to understand it).

Reply:
Alternatively, it would still be clearer if `ctx` had an `updateRequest` method which contained this `instanceof` since it seems like an implementation detail of that class, not here.

Reply:
`RequestContext` already has a method `setRequest()` which raises an exception when the original request is of a different type than that of the given one. What do you think about adding `trySetRequest()`? Perhaps we could add some general explanation about when to use it rather than `setRequest()` to the Javadoc of `trySetRequest()`.

Reply:
Renamed `setRequest()` to `updateRequest()` and made it return `boolean` instead of raising an exception on type mismatch.